### PR TITLE
nt-txn: reduce memory usage by avoiding sort executor |tidb-test=release-6.5

### PR DESCRIPTION
**Pull request by you06**
Signed-off-by: you06 <you1474600@gmail.com>

### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

Non-transactional may generate sharding query like:

```
SELECT `_tidb_rowid` FROM `xx`.`yy` WHERE TRUE ORDER BY IF(ISNULL(`_tidb_rowid`),0,1),`_tidb_rowid`
```

In case `_tidb_rowid` is not indexed, the query will be executed by a sort executor, which consumes large memory.

### What is changed and how it works?

This PR avoids using the sort executor for such case and sorts the `_tidb_rowid` by quick sort, in my test, 1,000,000,000 rand int64 numbers(worst case) take 5m56s to sort with 8G memory usage, this definitely outperforms the sort executor.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: session/nontransactional.go [hunk 1]
```diff
@@ -759,37 +768,138 @@ func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDMLStmt, se S
...
+			return nil, err
+		}
+		// last chunk
+		if chk.NumRows() == 0 {
+			break
+		}
+		iter := chunk.NewIterator4Chunk(chk)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			current := row.GetDatum(0, &rs.Fields()[0].Column.FieldType)
+			rowIDs = append(rowIDs, current.GetInt64())
```
Comment (ekexium) [inline session/nontransactional.go:818]: Storing all rowIDs can still result in uncontrolled memory usage.

File: session/nontransactional.go [hunk 2]
```diff
@@ -759,37 +768,138 @@ func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDMLStmt, se S
...
+	sort.Sort(Int64Slice(rowIDs))
+
+	jobCount := 0
+	currentSize := 0
+	jobs := make([]job, 0, (len(rowIDs)/batchSize)+2)
+	if len(rowIDs) > 0 {
+		var start *types.Datum
+		for _, i := range rowIDs {
+			if start == nil {
+				start = &types.Datum{}
```
Comment (ekexium) [inline session/nontransactional.go:831]: `NewIntDatum()`?

File: session/nontransactional.go [hunk 3]
```diff
@@ -759,37 +768,138 @@ func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDMLStmt, se S
...
+	// _tidb_rowid may be null in join, handle this case
+	if stmt.TransferFromInsert {
+		var nullStart, nullEnd types.Datum
+		nullStart.SetNull()
+		nullEnd.SetNull()
+		jobs = appendNewJob(jobs, 0, nullStart, nullEnd, 1, memTracker)
+		if len(jobs) > 1 {
+			lastIdx := len(jobs) - 1
+			jobs[0], jobs[lastIdx] = jobs[lastIdx], jobs[0]
+		}
```
Comment (ekexium) [inline session/nontransactional.go:860]: Why is it needed to swap?

**Review by ekexium — APPROVED** [review]
LGTM as a hotfix.
But I think reading `_tidb_rowid` in order should't involve any actual sorting work. After all it is stored in order.

File: session/nontransactional.go (hunk 2 above)
Comment (you06) [reply session/nontransactional.go:831]: NewIntDatum returns `types.Datum` but we need a pointer here.

File: session/nontransactional.go (hunk 3 above)
Comment (you06) [reply session/nontransactional.go:860]: Here we handle the case `_tidb_rowis IS NULL`, I'm afraid there are too many NULL values, put it to the head to reduce the impact of the possible slow execution.

File: session/nontransactional.go (hunk 1 above)
Comment (you06) [reply session/nontransactional.go:818]: The available memory is enough for now.

**Comment by you06** [comment]
> But I think reading `_tidb_rowid` in order should't involve any actual sorting work. After all it is stored in order.

Yes, but the executor doesn't guarantee it's read in order.

**Review by cfzjywxk — APPROVED** [review]

**Review by ekexium — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**